File: creational_patterns/go/prototype/order.go
/**
 * @file order.go
 * @author tushanFox (tushanFox@outlook.com)
 * @brief 创建者模式，以订单为例。
 * @version 0.1
 * @date 2022-02-14
 *
 * @copyright Copyright (c) 2022
 *
 */
package main

import "fmt"

// -----------------------------------------------------------------
// 抽象原型
type OrderIntf interface {
	SetOrderId(id string)
	GetOrderId() string
	SetProductNum(num int)
	GetProductNum() int
	ShowOrder()
	Clone() OrderIntf
}

// -----------------------------------------------------------------
// 具体原型
// 个人订单
type PersonalOrder struct {
	customerName string
	orderId      string
	productNum   int
}

func (p *PersonalOrder) SetOrderId(id string) {
	p.orderId = id
}
func (p *PersonalOrder) GetOrderId() string {
	return p.orderId
}
func (p *PersonalOrder) SetProductNum(num int) {
	p.productNum = num
}
func (p *PersonalOrder) GetProductNum() int {
	return p.productNum
}
func (p *PersonalOrder) SetCustomerName(name string) {
	p.customerName = name
}
func (p *PersonalOrder) GetCustomerName() string {
	return p.customerName
}
func (p *PersonalOrder) ShowOrder() {
	fmt.Println(p.orderId, "|", p.productNum, "|", p.customerName)
}
func (p *PersonalOrder) Clone() OrderIntf {
	fmt.Println("cloning PersonalOrder")
	order := &PersonalOrder{
		customerName: p.customerName,
		orderId:      p.orderId,
		productNum:   p.productNum,
	}
	return order
}

// -----------------------------------------------------------------
// 具体原型
// 公司订单
type EnterpriseOrder struct {
	enterpriseName string
	orderId        string
	productNum     int
}

func (p *EnterpriseOrder) SetOrderId(id string) {
	p.orderId = id
}
func (p *EnterpriseOrder) GetOrderId() string {
	return p.orderId
}
func (p *EnterpriseOrder) SetProductNum(num int) {
	p.productNum = num
}
func (p *EnterpriseOrder) GetProductNum() int {
	return p.productNum
}
func (p *EnterpriseOrder) SetEnterpriseName(name string) {
	p.enterpriseName = name
}
func (p *EnterpriseOrder) GetEnterpriseName() string {
	return p.enterpriseName
}
func (p *EnterpriseOrder) ShowOrder() {
	fmt.Println(p.orderId, "|", p.productNum, "|", p.enterpriseName)
}
func (p *EnterpriseOrder) Clone() OrderIntf {
	fmt.Println("cloning EnterpriseOrder")
	order := &EnterpriseOrder{
		enterpriseName: p.enterpriseName,
		orderId:        p.orderId,
		productNum:     p.productNum,
	}
	return order
}

File: creational_patterns/cpp/prototype/order.cpp
/**
 * @file order.cpp
 * @author tushanFox (tushanFox@outlook.com)
 * @brief 原型模式，以订单举例。
 * @version 0.1
 * @date 2022-01-14
 *
 * @copyright Copyright (c) 2022
 *
 */
#include "order.h"
#include <iostream>
// -----------------------------------------------------------------
// 具体原型类
// 个人订单
OrderIntf *PersonalOrder::Clone() {
  PersonalOrder *order = new PersonalOrder();
  std::cout << "cloning PersonalOrder" << std::endl;
  order->SetOrderId(order_id_);
  order->SetProductNum(product_num_);
  order->SetCustomerName(customer_name_);
  return order;
}
void PersonalOrder::SetOrderId(const std::string &id) {
  order_id_ = id;
}
std::string &PersonalOrder::GetOrderId() {
  return order_id_;
}
void PersonalOrder::SetProductNum(int num) {
  product_num_ = num;
}
int PersonalOrder::GetProductNum() {
  return product_num_;
}
void PersonalOrder::SetCustomerName(const std::string &name) {
  customer_name_ = name;
}
std::string &PersonalOrder::GetCustomerName() {
  return customer_name_;
}
void PersonalOrder::ShowOrder() {
  std::cout << order_id_ << "|" << product_num_ << "|" << customer_name_ << std::endl;
}

// -----------------------------------------------------------------
// 具体原型类
// 公司订单
OrderIntf *EnterpriseOrder::Clone() {
  EnterpriseOrder *order = new EnterpriseOrder();
  std::cout << "cloning EnterpriseOrder" << std::endl;
  order->SetOrderId(order_id_);
  order->SetProductNum(product_num_);
  order->SetEnterpriseName(enterprise_name_);
  return order;
}
void EnterpriseOrder::SetOrderId(const std::string &id) {
  order_id_ = id;
}
std::string &EnterpriseOrder::GetOrderId() {
  return order_id_;
}
void EnterpriseOrder::SetProductNum(int num) {
  product_num_ = num;
}
int EnterpriseOrder::GetProductNum() {
  return product_num_;
}
void EnterpriseOrder::SetEnterpriseName(const std::string &name) {
  enterprise_name_ = name;
}
std::string &EnterpriseOrder::GetEnterpriseName() {
  return enterprise_name_;
}
void EnterpriseOrder::ShowOrder() {
  std::cout << order_id_ << "|" << product_num_ << "|" << enterprise_name_ << std::endl;
}